File: src/main/java/au/id/micolous/metrodroid/transit/Transaction.java
/*
 * TransactionTrip.java
 *
 * Copyright 2018 Google
 * Copyright 2018-2019 Michael Farrell <micolous+git@gmail.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package au.id.micolous.metrodroid.transit;

import android.os.Parcelable;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import java.util.Calendar;
import java.util.Collections;
import java.util.List;

public abstract class Transaction implements Parcelable, Comparable<Transaction> {
    protected abstract boolean isTapOff();

    /**
     * This method may be overridden to provide candidate line names associated with the
     * transaction. This is useful if there is a separate field on the card which encodes the route
     * or line taken, and that knowledge of the station alone is not generally sufficient to
     * determine the correct route.
     *
     * By default, this gets candidate route names from the Station.
     */
    @NonNull
    public List<String> getRouteNames() {
        Station s = getStation();
        return s != null ? s.getLineNames() : Collections.emptyList();
    }

    /**
     * This method may be overridden to provide candidate line names associated with the
     * transaction. This is useful if there is a separate field on the card which encodes the route
     * or line taken, and that knowledge of the station alone is not generally sufficient to
     * determine the correct route.
     *
     * By default, this gets candidate route names from the Station.
     */
    @NonNull
    public List<String> getHumanReadableLineIDs() {
        Station s = getStation();
        return s != null ? s.getHumanReadableLineIDs() : Collections.emptyList();
    }

    public String getVehicleID() {
        return null;
    }

    public String getMachineID() { return null; }

    public int getPassengerCount() {
        return -1;
    }

    @Nullable
    public String getAgencyName(boolean isShort) {
        return null;
    }

    @Nullable
    public Station getStation() {
        return null;
    }

    public abstract Calendar getTimestamp();

    public abstract TransitCurrency getFare();

    public Trip.Mode getMode() {
        return Trip.Mode.OTHER;
    }

    protected boolean shouldBeMerged(@NonNull Transaction other) {
        return isTapOn() && (other.isTapOff() || other.isCancel()) && isSameTrip(other);
    }

    protected boolean isCancel() {
        return false;
    }

    protected abstract boolean isSameTrip(@NonNull Transaction other);

    protected abstract boolean isTapOn();

    protected boolean isTransfer() {
        return false;
    }

    protected boolean isRejected() {
        return false;
    }

    public int compareTo(@NonNull Transaction other) {
        final Calendar t1 = getTimestamp();
        final Calendar t2 = other.getTimestamp();
        if (t1 != null && t2 != null) {
            return t1.compareTo(t2);
        } else if (t2 != null) {
            return 1;
        } else {
            return 0;
        }
    }

    public static class Comparator implements java.util.Comparator<Transaction> {
        @Override
        public int compare(@NonNull Transaction txn1, @NonNull Transaction txn2) {
            return txn1.compareTo(txn2);
        }
    }
}

File: src/main/java/au/id/micolous/metrodroid/transit/Transaction.kt
/*
 * TransactionTrip.java
 *
 * Copyright 2018 Google
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package au.id.micolous.metrodroid.transit

import android.os.Parcelable

import java.util.Calendar

abstract class Transaction : Parcelable, Comparable<Transaction> {
    protected abstract val isTapOff: Boolean

    /**
     * This method may be overridden to provide candidate line names associated with the
     * transaction. This is useful if there is a separate field on the card which encodes the route
     * or line taken, and that knowledge of the station alone is not generally sufficient to
     * determine the correct route.
     *
     * By default, this gets candidate route names from the Station.
     */
    open val routeNames: List<String>
        get() {
            val s = station
            return s?.lineNames ?: emptyList()
        }

    /**
     * This method may be overridden to provide candidate line names associated with the
     * transaction. This is useful if there is a separate field on the card which encodes the route
     * or line taken, and that knowledge of the station alone is not generally sufficient to
     * determine the correct route.
     *
     * By default, this gets candidate route names from the Station.
     */
    open val humanReadableLineIDs: List<String>
        get() {
            val s = station
            return s?.humanReadableLineIDs ?: emptyList()
        }

    open val vehicleID: String?
        get() = null

    open val machineID: String?
        get() = null

    open val passengerCount: Int
        get() = -1

    open val station: Station?
        get() = null

    abstract val timestamp: Calendar?

    abstract val fare: TransitCurrency

    open val mode: Trip.Mode
        get() = Trip.Mode.OTHER

    protected open val isCancel: Boolean
        get() = false

    protected abstract val isTapOn: Boolean

    protected open val isTransfer: Boolean
        get() = false

    protected open val isRejected: Boolean
        get() = false

    open fun getAgencyName(isShort: Boolean): String? {
        return null
    }

    protected open fun shouldBeMerged(other: Transaction): Boolean {
        return isTapOn && (other.isTapOff || other.isCancel) && isSameTrip(other)
    }

    protected abstract fun isSameTrip(other: Transaction): Boolean

    override fun compareTo(other: Transaction): Int {
        val t1 = timestamp
        val t2 = other.timestamp
        if (t1 == null && t2 == null)
            return 0
        if (t1 == null)
            return -1
        if (t2 == null)
            return +1
        return t1.compareTo(t2)
    }

    class Comparator : java.util.Comparator<Transaction> {
        override fun compare(txn1: Transaction, txn2: Transaction): Int {
            return txn1.compareTo(txn2)
        }
    }
}